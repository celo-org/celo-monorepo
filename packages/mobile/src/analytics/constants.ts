import { DefaultEventNames } from '@celo/react-components/analytics/constants'

export enum CustomEventNames {
  language_select = 'language_select',
  nux_continue = 'nux_continue',
  full_name_input = 'full_name_input',
  invitation_code = 'invitation_code',
  signup_submit = 'signup_submit',
  blockChainCorruption = 'block_chain_corruption',
  disconnect_banner = 'disconnect_banner',
  gold_cancel1 = 'gold_cancel1',
  gold_cancel2 = 'gold_cancel2',
  gold_cancel3 = 'gold_cancel3',
  photo_education_cancel1 = 'photo_education_cancel1',
  photo_education_cancel2 = 'photo_education_cancel2',
  photo_education_cancel3 = 'photo_education_cancel3',
  exchange_gold_nux = 'exchange_gold_nux',
  wallet_gold_nux = 'wallet_gold_nux',
  dollar_cancel1 = 'dollar_cancel1',
  dollar_cancel2 = 'dollar_cancel2',
  dollar_cancel3 = 'dollar_cancel3',
  send_dollar_nux = 'send_dollar_nux',
  wallet_dollar_nux = 'wallet_dollar_nux',
  send_input = 'send_input',
  transaction_details = 'transaction_details',
  send_continue = 'send_continue',
  send_select_cancel = 'send_select_cancel',
  send_dollar_confirm = 'send_dollar_confirm',
  edit_dollar_confirm = 'edit_dollar_confirm',
  send_invite_details = 'send_invite_details',
  send_invite = 'send_invite',
  edit_send_invite = 'edit_send_invite',

  // Send events, separate from button tracking above
  send_dollar_transaction = 'send_dollar_transaction',
  send_dollar_transaction_confirmed = 'send_dollar_transaction_confirmed',

  fetch_balance = 'fetch_balance',

  // Verification event and sub-events
  verification = 'verification',
  verification_setup = 'verification_setup',
  verification_get_status = 'verification_get_status',
  verification_request_attestations = 'verification_request_attestations',
  verification_requested_attestations = 'verification_requested_attestations',
  verification_get_attestations = 'verification_get_attestations',
  verification_set_account = 'verification_set_account',
  verification_reveal_attestation = 'verification_reveal_attestation',
  verification_revealed_attestation = 'verification_revealed_attestation',
  verification_codes_received = 'verification_codes_received',
  verification_complete_attestation = 'verification_complete_attestation',
  verification_completed_attestation = 'verification_completed_attestation',
  verification_manual_selected = 'verification_manual_selected',
  verification_failed = 'verification_failed',
  verification_cancelled = 'verification_cancelled',
  verification_success = 'verification_success',
  verification_timed_out = 'verification_timed_out',

  verification_actionable_attestation_start = 'verification_actionable_attestation_start',
  verification_actionable_attestation_finish = 'verification_actionable_attestation_finish',
  verification_validate_code_start = 'verification_validate_code_start',
  verification_validate_code_finish = 'verification_validate_code_finish',

  redeem_invite_success = 'redeem_invite_success',
  redeem_invite_timed_out = 'redeem_invite_timed_out',
  redeem_invite_failed = 'redeem_invite_failed',

  photos_education = 'photos_education',
  get_backup_key = 'earn_celo_gold',
  earn_celo_gold = 'earn_celo_gold',
  user_restart = 'user_restart',
  pin_continue = 'pin_continue',
  pin_wallet_import = 'pin_wallet_import',
  pin_value = 'pin_value',
  pin_create_button = 'pin_create_button',

  // Screen name: Invite_Friends, Invite_Friends_Fee
  friend_invited = 'friend_invited', // to record the name or number of the friend a user is inviting per an invite session
  // TODO: this event is tracked, but isn't showing up on screen, so it is untested
  invite_cancel = 'invite_cancel', // to count the # of taps on the “Cancel” element in the upper left corner
  // TODO: Add this metric later. There doens't appear to be a Cancel button in the upper left of Invite_Friends_Review
  // invite_fee_cancel = 'invite_fee_cancel', // to count the # of taps on “Cancel” in upper left of Invite_Friends_Review
  invite_edit = 'invite_edit', // to count the # of taps on “Cancel” button at bottom of Invite_Friends_Review
  invite_friends_sms = 'invite_friends_sms', // to count the # of taps on “Invite with SMS" button on Invite_Friends_Review
  invite_friends_whatsapp = 'invite_friends_whatsapp', // to count the # of taps on “Invite with WhatsApp" button on Invite_Friends_Review

  // Screen name: Backup_Phrase, Backup_Insist, Backup_Share, Backup_Set
  set_backup_phrase = 'set_backup_phrase', // (count # of taps on “Set Backup Phrase” in Backup_Phrase) [we should not track the actual value of this field, just whether the user filled it out]
  set_social_backup = 'set_social_backup', // (count # of taps on "Set up Social Backup")
  delay_backup = 'delay_backup', // (Count # of taps on "Delay" button in Backup_Phrase)
  skip_backup = 'skip_backup', // (count # of taps on “Skip” button in Backup_Phrase)
  view_backup_phrase = 'view_backup_phrase', // (count # of taps on "View Backup Phrase" after already backed up)
  view_social_backup = 'view_social_backup', // (count # of taps on "View Social Backup" after already set up)
  skip_social_backup = 'skip_social_backup', // (count # of taps on "Skip Social Backup" )
  backup_cancel = 'backup_cancel', // (count # of taps on "Cancel" button in Backup_Phrase)
  insist_backup_phrase = 'insist_backup_phrase', // (count # of taps on “Set Backup Phrase” in Backup_Insist)
  insist_skip_backup = 'insist_skip_backup', // (count # of taps on “Do Later” in Backup_Insist)
  whatsapp_backup = 'whatsapp_backup', // (count # of taps on “Send with Whatsapp” in Backup_Share)
  backup_continue = 'backup_continue', // (count # of taps on “Continue” button in Backup_Phrase)
  social_backup_continue = 'social_backup_continue', // (Count # of taps on "Backup with Friends" in Backup_Phrase)

  // Screen name: Backup_Quiz, Question_Incorrect, Backup_Confirmed
  question_select = 'question_select', // (track # of input selections on Backup_Verify screen)
  question_submit = 'question_submit', // (track # of taps on “Submit” button for Backup_Quiz screen)
  question_cancel = 'questions_cancel', // (track # of taps on "Cancel" button on the Backup_Quiz Screens)
  question_incorrect = 'question_incorrect', // (track # of taps on “See Backup Phrase” in Backup_Quiz)
  question_done = 'question_done', // (track # of taps on “Done” button on the Backup_Confirmed screen)

  // Screen name: Backup_Verify
  backup_paste = 'backup_paste', // (track # of pastes in input field for Backup_Verify screen)
  backup_paste_submit = 'backup_paste_submit', // (track # of taps on "Submit" button for Backup_Verify screen)
  backup_paste_cancel = 'backup_paste_cancel', // (track # of taps on "Cancel" button on the Backup_Verify screen)
  backup_paste_incorrect = 'backup_paste_incorrect', // (track # of taps on "See Backup Phrase" in Backup_Verify screen)
  backup_paste_done = 'backup_paste_done', // (track # of taps on "Done" button on the Backup_Verify screen)

  // Screens: Exchange_Tutorial, Exchange_Home, Exchange_Currency
  exchange_button = 'exchange_button', // count # of taps on the exchange button in Exchange_Home
  exchange_dollar_input = 'exchange_dollar_input', // record the input of Celo$ form field in Exchange_Currency
  // triggered when focus is moved away from text editing
  exchange_gold_input = 'exchange_gold_input', // record the input of Gold form field in Exchange_Currency,
  // triggered when focus is moved away from text editing
  exchange_continue = 'exchange_continue', // record # of taps on “Continue” button on Exchange_Currency
  currency_swap = 'currency_swap', // record # of taps on the double arrow button in Exchange_Currency

  // Screen: Exchange_Review
  exchange_confirm = 'exchange_confirm', // to count the # of taps on the exchange button here)
  exchange_edit = 'exchange_edit', // to count # of times users click the edit button, change transaction)
  exchange_cancel = 'exchange_cancel', // to count # of taps on the “Cancel” in the upper left corner)

  // Screen name: Account_Home, Account_Edit, Account_Edit_Name
  edit_profile = 'edit_profile', // to count the # of taps on the edit profile button in Account_Home
  edit_name = 'edit_name', // to count the # of taps on the “Edit Name” nav element in Account_Edit
  edit_name_input = 'edit_name_input', // to record the input of the form input field on Account_Edit_Name page
  edit_name_submit = 'edit_name_submit', // to count the # of taps on the "Done" button on Account_Edit_Name page
  edit_account_cancel = 'edit_account_cancel', // to count the # of taps on “Cancel” in upper left of Account_Edit
  edit_name_cancel = 'edit_name_cancel', // to count the # of taps on “Cancel” in upper left of Account_Edit_Name

  // Screen names: Wallet_Recover, Wallet_Import
  import_phrase_input = 'import_phrase_input', // to record the # of times a value is inputted here [we should not track the actual value of this field, just whether the user filled it out]
  import_wallet_submit = 'import_wallet_submit', // to count the # of times that the “Restore Celo Wallet” button is pressed
  import_wallet_cancel = 'import_wallet_cancel', // to count the # of times that the “Cancel” button is pressed

  // Request Money events
  // request_select_cancel is captured in send_select_cancel
  request_payment_continue = 'incoming_request_payment_continue',
  request_payment_request = 'incoming_request_payment_request',
  request_payment_edit = 'incoming_request_payment_edit',

  incoming_request_payment_decline = 'incoming_request_payment_decline',
  incoming_request_payment_pay = 'incoming_request_payment_pay',
  incoming_request_payment_review = 'incoming_request_payment_review',

  outgoing_request_payment_review = 'outgoing_request_payment_review',
  outgoing_request_payment_remind = 'outgoing_request_payment_remind',
  outgoing_request_payment_cancel = 'outgoing_request_payment_cancel',

  // Escrowed payments
  escrowed_payment_review = 'escrowed_payment_review',
  escrowed_payment_reclaimed_by_sender = 'escrowed_payment_reclaimed_by_sender',
  escrowed_payment_reclaimEdit_by_sender = 'escrowed_payment_reclaimEdit_by_sender',
  escrowed_payment_withdrawn_by_receiver = 'escrowed_payment_withdrawn_by_receiver',
  clicked_escrowed_payment_notification = 'clicked_escrowed_payment_notification',
  clicked_escrowed_payment_send_message = 'clicked_escrowed_payment_send_message',

  // Notifications
  celogold_notification_confirm = 'celogold_notification_confirm',
  celogold_notification_dismiss = 'celogold_notification_dismiss',
  celorewards_notification_confirm = 'celorewards_notification_confirm',
  celorewards_notification_dismiss = 'celorewards_notification_dismiss',
  invitefriends_notification_confirm = 'invitefriends_notification_confirm',
  invitefriends_notification_dismiss = 'invitefriends_notification_dismiss',

  // QR Code
  qrcode_main_screen_visit = 'qrcode_main_screen_visit',

  // Performance
  transaction_send_start = 'transaction_send_start',
  transaction_send_gas_estimated = 'transaction_send_gas_estimated',
  transaction_send_gas_hash_received = 'transaction_send_gas_hash_received',
  transaction_send_gas_receipt = 'transaction_send_gas_receipt',
}

export enum CommonValues {
  success = 'success',
  failure = 'failure',
  cancel = 'cancel',
  timeout = 'timeout',
}

// TODO(nitya): separate this out by event name
export const PROPERTY_PATH_WHITELIST = [
  'address',
  'component',
  'countryCode',
  'cta',
  'currentScreen',
  'dollarBalance',
  'dollarPendingBalance',
  'error',
  'exchangeInputAmount',
  'exchangeRate',
  'fullName',
  'goldBalance',
  'goldPendingBalance',
  'inviteCode',
  'isCorrect',
  'issuer',
  'label',
  'language',
  'makerAmount',
  'makerToken',
  'makerTokenAmount',
  'name',
  'navigation.state.key',
  'navigation.state.routeName',
  'nextScreen',
  'note',
  'phoneNumber',
  'previousScreen',
  'query',
  'recipientAddress',
  'requesteeAddress',
  'requestIndex',
  'rootTag',
  'routeName',
  'screen',
  'selectedAnswer',
  'selectedRecipientAddress',
  'selectedRecipientPhoneNumber',
  'sendAmount',
  'subtitle',
  'success',
  'syncProgress',
  'takerAmount',
  'testnet',
  'timeElapsed',
  'title',
  'tti',
  'txId',
  'verificationIndex',
  'verificationsRemaining',
]
export { DefaultEventNames }
