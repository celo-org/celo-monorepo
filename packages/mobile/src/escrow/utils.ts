import { ensureLeading0x } from '@celo/base'
import { generateDeterministicInviteCode } from '@celo/utils/lib/account'
import { publicKeyToAddress } from '@celo/utils/lib/address'

export const generateEscrowPaymentIdAndPk = (
  recipientPhoneHash: string,
  recipientPepper: string,
  addressIndex: number = 0
) => {
  const { publicKey, privateKey } = generateDeterministicInviteCode(
    recipientPhoneHash,
    recipientPepper,
    addressIndex
  )
  return { paymentId: publicKeyToAddress(publicKey), privateKey: ensureLeading0x(privateKey) }
}

export const generateUniquePaymentId = (
  existingPaymentIds: string[],
  phoneHash: string,
  pepper: string
) => {
  const paymentIdSet: Set<string> = new Set(existingPaymentIds)

  // Using an upper bound of 100 to be sure this doesn't run forever given
  // the realistic amount of pending escrow txs is far less than this
  for (let i = 0; i < 100; i += 1) {
    const { paymentId } = generateEscrowPaymentIdAndPk(phoneHash, pepper, i)
    if (!paymentIdSet.has(paymentId)) {
      return paymentId
    }
  }
}
