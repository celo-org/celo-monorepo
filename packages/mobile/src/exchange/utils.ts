import BigNumber from 'bignumber.js'
import { CURRENCY_ENUM } from 'src/geth/consts'
import { convertToContractDecimals } from 'src/tokens/saga'
import { getContractKitAsync } from 'src/web3/contracts'

export async function celoTocUsd(celoAmount: BigNumber) {
  const celoAmountInWei = (
    await convertToContractDecimals(celoAmount, CURRENCY_ENUM.GOLD)
  ).integerValue()

  const contractKit = await getContractKitAsync()
  const exchange = await contractKit.contracts.getExchange()
  const exchangeRate = await exchange.getGoldExchangeRate(celoAmountInWei)
  const cUsdAmount = exchangeRate.multipliedBy(celoAmount)

  return cUsdAmount
}
