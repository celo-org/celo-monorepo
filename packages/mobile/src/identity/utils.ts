import BigNumber from 'bignumber.js'

export function isUserBalanceSufficient(
  userBalance: string | number | null,
  estimatedTxAmount: number | string
) {
  if (!userBalance || new BigNumber(userBalance).isLessThan(new BigNumber(estimatedTxAmount))) {
    return false
  }

  return true
}
