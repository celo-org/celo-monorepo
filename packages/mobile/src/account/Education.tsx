import Button, { BtnTypes } from '@celo/react-components/components/Button'
import Touchable from '@celo/react-components/components/Touchable'
import colors from '@celo/react-components/styles/colors'
import { fontStyles } from '@celo/react-components/styles/fonts'
import * as React from 'react'
import { WithNamespaces, withNamespaces } from 'react-i18next'
import { Dimensions, Image, StyleSheet, Text, View } from 'react-native'
import Swiper from 'react-native-swiper'
import CeloAnalytics from 'src/analytics/CeloAnalytics'
import { CustomEventNames } from 'src/analytics/constants'
import { componentWithAnalytics } from 'src/analytics/wrapper'
import { Namespaces } from 'src/i18n'
import { placeholder } from 'src/images/Images'
import { navigateBack } from 'src/navigator/NavigationService'

const PROGRESS_CIRCLE_PASSIVE_SIZE = 8
const PROGRESS_CIRCLE_ACTIVE_SIZE = 12

export const CTA_CIRCLE_SIZE = 5

interface State {
  step: number
}

interface EducationStep {
  image: any
  text: string
  cancelEvent: CustomEventNames
  screenName: string
}

interface CustomizedProps {
  stepInfo: EducationStep[]
  buttonText: string
  linkText?: string
  onFinish: () => void
  onFinishAlternate?: () => void
}

type Props = WithNamespaces & CustomizedProps

class Education extends React.Component<Props, State> {
  state = {
    step: 0,
  }

  goBack = () => {
    const currentStepInfo = this.props.stepInfo[this.state.step]
    if (currentStepInfo.cancelEvent && currentStepInfo.screenName) {
      CeloAnalytics.track(currentStepInfo.cancelEvent, {
        screen: currentStepInfo.screenName,
      })
    }
    navigateBack()
  }

  setStep = (step: number) => {
    this.setState({ step })
  }

  renderHeader() {
    const { t } = this.props
    return (
      <View style={style.header}>
        <View style={style.goBack}>
          <Touchable testID="Education-goback" borderless={true} onPress={this.goBack}>
            <Text style={fontStyles.headerButton}> {t('cancel')}</Text>
          </Touchable>
        </View>
      </View>
    )
  }

  renderBody() {
    const { t, stepInfo } = this.props
    const children = stepInfo.map((v: any, i: any) => {
      const imgSrc = v.image ? v.image : placeholder
      return (
        <View style={style.swipedContent} key={i}>
          <Image source={imgSrc} style={style.bodyImage} resizeMode="contain" />
          <Text style={[fontStyles.h1, style.bodyText]}>{t(v.text)}</Text>
        </View>
      )
    })

    return (
      <View style={style.body}>
        {/* 
        // @ts-ignore */}
        <Swiper
          onIndexChanged={this.setStep}
          loop={false}
          showsButtons={false}
          showsPagination={true}
          style={style.swiper}
          dotStyle={style.circlePassive}
          activeDotStyle={style.circleActive}
          containerStyle={style.swiperContainer}
        >
          {children}
        </Swiper>
      </View>
    )
  }

  renderFooter() {
    const { t, stepInfo, onFinish, onFinishAlternate, buttonText, linkText } = this.props
    if (this.state.step !== stepInfo.length - 1) {
      return <View style={style.footer} />
    }
    return (
      <View style={style.footer}>
        <View style={style.buttonContainer}>
          <Button
            text={t(buttonText)}
            onPress={onFinish}
            style={style.button}
            standard={true}
            type={BtnTypes.PRIMARY}
          />
        </View>
        {onFinishAlternate ? (
          <Button
            text={t(linkText ? linkText : 'backToWallet')}
            onPress={onFinishAlternate}
            style={style.button}
            standard={true}
            type={BtnTypes.TERTIARY}
          />
        ) : null}
      </View>
    )
  }

  render() {
    return (
      <View style={style.container}>
        {this.renderHeader()}
        {this.renderBody()}
        {this.renderFooter()}
      </View>
    )
  }
}

const circle = {
  flex: 0,
  backgroundColor: colors.inactive,
  borderRadius: 8,
}

const { width } = Dimensions.get('window')
const style = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'white',
  },
  header: {
    padding: 20,
    margin: 0,
    flexDirection: 'row',
    justifyContent: 'flex-start',
    width: '100%',
  },
  goBack: {
    flex: 0,
  },
  body: {
    flex: 8,
    alignItems: 'center',
  },
  bodyText: {
    color: colors.darkSecondary,
    paddingTop: 30,
    paddingHorizontal: 20,
  },
  bodyImage: {
    alignSelf: 'center',
    width: 200,
    height: 200,
  },
  footer: {
    justifyContent: 'flex-end',
    alignSelf: 'stretch',
    flex: 2,
  },
  buttonContainer: {
    flexDirection: 'row',
  },
  button: {
    flex: 1,
    paddingHorizontal: 20,
  },
  circle,
  circlePassive: {
    ...circle,
    height: PROGRESS_CIRCLE_PASSIVE_SIZE,
    width: PROGRESS_CIRCLE_PASSIVE_SIZE,
  },
  circleActive: {
    ...circle,
    height: PROGRESS_CIRCLE_ACTIVE_SIZE,
    width: PROGRESS_CIRCLE_ACTIVE_SIZE,
  },
  swiper: {
    alignItems: 'center',
    flex: 1,
    width,
  },
  swiperContainer: {
    alignItems: 'center',
  },
  swipedContent: {
    flex: 1,
    justifyContent: 'center',
    marginBottom: 60,
  },
})

export default componentWithAnalytics(withNamespaces(Namespaces.nuxCurrencyPhoto4)(Education))
