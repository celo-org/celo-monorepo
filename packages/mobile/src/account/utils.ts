import { ADDRESS_LENGTH } from 'src/exchange/reducer'

export const isAddressFormat = (content: string): boolean => {
  return content.startsWith('0x') && content.length === ADDRESS_LENGTH
}
