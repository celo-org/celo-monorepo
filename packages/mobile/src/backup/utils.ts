import { sampleSize } from 'lodash'
import { AsyncStorage } from 'react-native'
import { generateMnemonic, wordlists } from 'react-native-bip39'
import { getKey, setKey } from 'src/utils/keyStore'
import Logger from 'src/utils/Logger'

const TAG = 'Backup/utils'

export async function createQuizWordList(mnemonic: string, language: string | null) {
  const disallowedWordSet = new Set(mnemonic.split(' '))
  const languageWordList = getWordlist(language)
  const wordOptions: string = await generateMnemonic(1000, null, languageWordList)
  const quizWordList = new Set(
    [...wordOptions.split(' ')].filter((word: string) => !disallowedWordSet.has(word))
  )
  return [...quizWordList]
}

export function selectQuizWordOptions(correctWord: string, allWords: string[], numOptions: number) {
  const wordOptions = []
  const correctWordPosition = Math.floor(Math.random() * numOptions)
  const randomWordIndexList = sampleSize([...Array(allWords.length).keys()], numOptions - 1)
  let randomWordIndex: number = 0

  for (let i = 0; i < numOptions; i++) {
    if (i === correctWordPosition) {
      wordOptions.push(correctWord)
      continue
    }

    wordOptions.push(allWords[randomWordIndexList[randomWordIndex]])
    randomWordIndex += 1
  }
  return wordOptions
}

export function getWordlist(language: string | null) {
  if (!language) {
    return wordlists.EN
  }

  switch (language.slice(0, 2)) {
    case 'es': {
      return wordlists.ES
    }
    default: {
      return wordlists.EN
    }
  }
}

// TODO(Rossy) Remove after the next alfa testnet reset
export async function getStoredMnemonic(): Promise<string | null> {
  try {
    Logger.debug(TAG, 'Checking keystore for mnemonic')
    let mnemonic = await getKey('mnemonic')
    if (mnemonic) {
      return mnemonic
    }

    Logger.debug(TAG, 'Mnemonic not found in keystore, checking async storage')
    mnemonic = await AsyncStorage.getItem('mnemonic')
    if (mnemonic) {
      await setKey('mnemonic', mnemonic)
      await AsyncStorage.removeItem('mnemonic')
      return mnemonic
    }

    Logger.error(TAG, 'No mnemonic found')
    return null
  } catch (error) {
    Logger.error(TAG, 'Failed to retrieve mnemonic', error)
    return null
  }
}
