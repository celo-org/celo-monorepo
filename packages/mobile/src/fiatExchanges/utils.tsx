import { SIMPLEX_URI } from 'src/config'
import { CURRENCY_ENUM } from 'src/geth/consts'
import { LocalCurrencyCode } from 'src/localCurrency/consts'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { navigateToURI } from 'src/utils/linking'

export const openMoonpay = (currencyCode: LocalCurrencyCode, currencyToBuy: CURRENCY_ENUM) => {
  navigate(Screens.MoonPay, {
    localAmount: 0,
    currencyCode,
    currencyToBuy,
  })
}

export const openSimplex = (account: string | null) => {
  navigateToURI(`${SIMPLEX_URI}?address=${account}`)
}
