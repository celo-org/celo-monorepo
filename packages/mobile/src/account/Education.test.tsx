import * as React from 'react'
import 'react-native'
import { fireEvent, render } from 'react-native-testing-library'
import { Provider } from 'react-redux'
import * as renderer from 'react-test-renderer'
import Education from 'src/account/Education'
import { CustomEventNames } from 'src/analytics/constants'
import { navigateBack } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { createMockStore } from 'test/utils'

const BUTTON_TEXT = 'Test'

const educationProps = {
  stepInfo: [
    {
      image: './test.jpg',
      text: 'Step 1',
      cancelEvent: CustomEventNames.photo_education_cancel1,
      screenName: Screens.Debug,
    },
  ],
  buttonText: BUTTON_TEXT,
  onFinish: jest.fn(),
}

describe('Education', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <Provider store={createMockStore({})}>
        <Education {...educationProps} />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })

  it('finishes when pressed', () => {
    const edu = render(<Education {...educationProps} />)
    fireEvent.press(edu.getByProps({ text: BUTTON_TEXT }))
    expect(educationProps.onFinish).toBeCalled()
  })

  it('navigates back', () => {
    const edu = render(<Education {...educationProps} />)
    fireEvent.press(edu.getByTestId('Education-goback'))
    expect(navigateBack).toBeCalled()
  })
})
