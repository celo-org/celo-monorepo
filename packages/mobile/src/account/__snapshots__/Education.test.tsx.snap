// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Education renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "margin": 0,
        "padding": 20,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 0,
        }
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        nativeBackgroundAndroid={
          Object {
            "attribute": "selectableItemBackgroundBorderless",
            "type": "ThemeAttrAndroid",
          }
        }
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="Education-goback"
      >
        <Text
          style={
            Object {
              "color": "#42D689",
              "fontFamily": "Hind-SemiBold",
              "fontSize": 14,
              "lineHeight": 22,
            }
          }
        >
           
          cancel
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 8,
      }
    }
  >
    <Swiper
      activeDotStyle={
        Object {
          "backgroundColor": "#D1D5D8",
          "borderRadius": 8,
          "flex": 0,
          "height": 12,
          "width": 12,
        }
      }
      containerStyle={
        Object {
          "alignItems": "center",
        }
      }
      dotStyle={
        Object {
          "backgroundColor": "#D1D5D8",
          "borderRadius": 8,
          "flex": 0,
          "height": 8,
          "width": 8,
        }
      }
      loop={false}
      onIndexChanged={[Function]}
      showsButtons={false}
      showsPagination={true}
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "width": 750,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "justifyContent": "center",
            "marginBottom": 60,
          }
        }
      >
        <Image
          resizeMode="contain"
          source="./test.jpg"
          style={
            Object {
              "alignSelf": "center",
              "height": 200,
              "width": 200,
            }
          }
        />
        <Text
          style={
            Array [
              Object {
                "color": "#2E3338",
                "fontFamily": "Hind-Light",
                "fontSize": 22,
                "paddingBottom": 20,
                "textAlign": "center",
              },
              Object {
                "color": "#81868B",
                "paddingHorizontal": 20,
                "paddingTop": 30,
              },
            ]
          }
        >
          Step 1
        </Text>
      </View>
    </Swiper>
  </View>
  <View
    style={
      Object {
        "alignSelf": "stretch",
        "flex": 2,
        "justifyContent": "flex-end",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "borderRadius": 3,
              "flexDirection": "row",
            },
            Object {
              "marginVertical": 10,
            },
            Object {
              "marginBottom": 10,
            },
            Object {
              "flex": 1,
              "paddingHorizontal": 20,
            },
            Object {
              "backgroundColor": "#42D689",
            },
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          nativeBackgroundAndroid={
            Object {
              "attribute": "selectableItemBackground",
              "type": "ThemeAttrAndroid",
            }
          }
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
              },
              Object {
                "backgroundColor": "#42D689",
              },
              Object {
                "height": 50,
              },
              Object {
                "borderColor": "#42D689",
                "borderRadius": 3,
                "borderWidth": 2,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "Hind-SemiBold",
                    "fontSize": 16,
                  },
                  Object {
                    "color": "#FFFFFF",
                  },
                  Object {
                    "paddingLeft": 5,
                    "paddingRight": 5,
                  },
                ]
              }
            >
              Test
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
